File: src/Cleansia.Core.Domain/ServiceAreas/CityNameMatch.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

namespace Cleansia.Core.Domain.ServiceAreas;

/// <summary>
/// Whether a customer's typed or geocoded city is the serviced city a <c>ServiceCity</c> row names.
///
/// <para><b>Why an exact compare was not enough.</b> The gate used to be
/// <c>row.Name.ToLower() == city.Trim().ToLower()</c>, which refuses a booking for any spelling the
/// operator did not seed verbatim. That is not hypothetical: this repo's own seed puts a customer at
/// <c>'Plzen'</c> (<c>sql-scripts/seed/insert_addresses.sql:39</c>) while the serviced row is
/// <c>'Plzeň'</c> (<c>insert_seed_data.sql:358</c>), so that address cannot be booked. Reverse geocoders
/// and hand-typing both drop diacritics, and both append districts.</para>
///
/// <para><b>The rule is MONOTONE — it accepts a strict superset of the exact compare.</b> Nothing that
/// books today stops booking. That is what makes it safe to ship on its own.</para>
///
/// <para><b>The district strip is one-directional and that is load-bearing.</b> It is applied only to
/// the CUSTOMER's string, never to the row's name. So a row named <c>Praha 8</c> continues to serve
/// only Praha 8 and does not quietly become a claim over the whole city — an operator who seeded a
/// single district meant a single district.</para>
///
/// <para><b>Exonyms are data, not algorithm.</b> <c>Prague</c> matches nothing but a <c>Prague</c> row,
/// and <c>Прага</c> matches nothing at all until somebody seeds it. Folding those in code would mean
/// shipping a translation table that goes stale the first time a market opens; a row is the honest
/// place for it, and the admin surface already creates rows.</para>
/// </summary>
public static class CityNameMatch
{
    /// <summary>
    /// A trailing 1–2 digit district, optionally followed by a quarter after a dash —
    /// <c>Praha 8</c>, <c>Praha 4 - Chodov</c>, <c>Praha 5 – Smíchov</c>.
    ///
    /// <para>A dash with NO leading number is deliberately not matched. <c>Praha-západ</c> and
    /// <c>Brno-venkov</c> have that exact shape and are <i>okresy</i> — the rural districts AROUND the
    /// city, not parts of it. Stripping them would serve the countryside on the strength of a row for
    /// the city.</para>
    /// </summary>
    private static readonly Regex DistrictSuffix = new(
        @"^(?<base>\S.*?)\s+\d{1,2}(\s*[-–—]\s*\S.*)?$",
        RegexOptions.Compiled | RegexOptions.CultureInvariant);

    private static readonly Regex WhitespaceRun = new(@"\s+", RegexOptions.Compiled);

    public static bool Matches(string? servicedCityName, string? customerCity)
    {
        var row = Fold(servicedCityName);
        var city = Fold(customerCity);

        if (row.Length == 0 || city.Length == 0)
        {
            return false;
        }

        return row == city || row == StripDistrict(city);
    }

    /// <summary>Trim, strip diacritics, lowercase, and collapse internal whitespace runs to one space.</summary>
    private static string Fold(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return string.Empty;
        }

        var decomposed = value.Trim().Normalize(NormalizationForm.FormD);
        var builder = new StringBuilder(decomposed.Length);
        foreach (var character in decomposed)
        {
            if (CharUnicodeInfo.GetUnicodeCategory(character) != UnicodeCategory.NonSpacingMark)
            {
                builder.Append(character);
            }
        }

        var stripped = builder.ToString().Normalize(NormalizationForm.FormC).ToLowerInvariant();
        return WhitespaceRun.Replace(stripped, " ");
    }

    /// <summary>Never returns empty — a bare "8" keeps its own shape and simply matches nothing.</summary>
    private static string StripDistrict(string folded)
    {
        var match = DistrictSuffix.Match(folded);
        return match.Success ? match.Groups["base"].Value : folded;
    }
}

File: src/Cleansia.App/libs/shared/utils/src/city-name-match.ts
/**
 * Whether a customer's typed or geocoded city is the serviced city a `ServiceCity` row names.
 *
 * **A port of the server's `Cleansia.Core.Domain.ServiceAreas.CityNameMatch`, and it has to stay
 * one.** The server is the authority: `OrderAddressResolver` runs the rule through
 * `ServiceCityRepository.CityIsServicedAsync` and refuses the booking with `city.not_serviced`.
 * This copy exists only so the customer hears it at address-selection time instead of at payment.
 *
 * **The danger in a copy is being STRICTER than the server, not looser.** A client that refuses a
 * city the server would accept tells a paying customer we do not serve them when we do — which is
 * exactly what the exact string compare this replaces did to the repo's own seeded `Plzen` address
 * against its `Plzeň` row. Looser is survivable: the customer proceeds and the server refuses,
 * which is what happened before any client check existed.
 *
 * The same table is pinned four times — `city-name-match.spec.ts`,
 * `Cleansia.Tests/Domain/ServiceAreas/CityNameMatchTests.cs`,
 * `cleansia_android/core/.../CityNameMatchTest.kt` and
 * `cleansia_ios/CleansiaCore/Tests/.../CityNameMatchTests.swift`. Change a case in one, change it
 * in all four.
 */

/**
 * A trailing 1–2 digit district, optionally followed by a quarter after a dash — `Praha 8`,
 * `Praha 4 - Chodov`, `Praha 5 – Smíchov`.
 *
 * A dash with NO leading number is deliberately not matched: `Praha-západ` and `Brno-venkov` have
 * that exact shape and are *okresy* — the rural rings around those cities, not parts of them.
 *
 * A numbered group rather than the `(?<base>…)` the C# and Kotlin originals use: this lib compiles
 * its specs at `target: es2016`, and TypeScript rejects named groups below es2018. Same reason the
 * mark class below is an explicit range rather than `\p{Mn}`.
 */
const DISTRICT_SUFFIX = /^(\S.*?)\s+\d{1,2}(?:\s*[-–—]\s*\S.*)?$/;

/** Combining marks left behind by NFD — exact for every diacritic these five markets use. */
const COMBINING_MARKS = /[̀-ͯ]/g;

const WHITESPACE_RUN = /\s+/g;

/** Trim, strip diacritics, lowercase, and collapse internal whitespace runs to one space. */
function fold(value: string | null | undefined): string {
  const trimmed = (value ?? '').trim();
  if (trimmed.length === 0) {
    return '';
  }
  return trimmed
    .normalize('NFD')
    .replace(COMBINING_MARKS, '')
    .normalize('NFC')
    .toLowerCase()
    .replace(WHITESPACE_RUN, ' ');
}

/** Never returns empty — a bare `8` keeps its own shape and simply matches nothing. */
function stripDistrict(folded: string): string {
  return DISTRICT_SUFFIX.exec(folded)?.[1] ?? folded;
}

/**
 * The strip runs on the CUSTOMER's string only, never on the row's name. A row named `Praha 8`
 * therefore keeps serving only Praha 8 — an operator who seeded one district meant one district.
 */
export function cityNameMatches(
  servicedCityName: string | null | undefined,
  customerCity: string | null | undefined,
): boolean {
  const row = fold(servicedCityName);
  const city = fold(customerCity);
  if (row.length === 0 || city.length === 0) {
    return false;
  }
  return row === city || row === stripDistrict(city);
}

/** True when ANY serviced city matches — the question a screen actually asks. */
export function isCityServiced(
  servicedCityNames: readonly (string | null | undefined)[],
  customerCity: string | null | undefined,
): boolean {
  return servicedCityNames.some((name) => cityNameMatches(name, customerCity));
}
